plugin/utils/compiler_builtins.py

```diff
@@ -0,0 +1,197 @@
+                    compiler, std, language)
+                CompilerBuiltIns.__cache[cfg] = (defines, includes)
+            self._defines = defines
+            self._include_paths = includes
+
+
+    @property
+    def defines(self):
+        """The built-in defines provided by the compiler."""
+        return self._defines
+
+    @property
+    def include_paths(self):
+        """The list of built-in include paths used by the compiler."""
+        return self._include_paths
+
+    @property
+    def flags(self):
+        """
+        The list of built-in flags.
+
+        This property holds the combined list of built-in defines and
+        include paths of the compiler.
+        """
+        return self._defines + self._include_paths
+
+    def _guess_compiler(self, args):
+        compiler = None
+        std = None
+        if len(args) > 0:
+            compiler = args[0]
+        else:
+            self._debug("Got empty command line - cannot extract compiler")
+        if len(args) > 1:
+            for arg in args[1:]:
+                if arg.startswith("-std="):
+                    std = arg[5:]
+        return (compiler, std)
+
+    def _guess_language(self, compiler):
+        """
+        Try to guess the language based on the compiler.
+
+        This is required as we need to explicitly pass a language when asking
+        the compiler later for its default flags.
+        """
+        if compiler.endswith("++"):
+            return "c++"
+        else:
+            return "c"
+
+    def _get_default_flags(self, compiler, std, language):
+        import subprocess
+        import re
+
+        result = list()
+
+        args = [compiler, "-x", language]
+        if std is not None:
+            args += ['-std=' + std]
+        args += ["-dM", "-E", "-"]
+
+        try:
+            res = subprocess.Popen(
+                args,
+                stdin=subprocess.DEVNULL,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT
+            )
+            res.wait()
+            for line in res.stdout.read().decode().splitlines():
+                m = re.search(r'#define ([\w()]+) (.+)', line)
+                if m is not None:
+                    result.append("-D{}={}".format(m.group(1), m.group(2)))
+                else:
+                    m = re.search(r'#define (\w+)', line)
+                    if m is not None:
+                        result.append("-D{}".format(m.group(1)))
+        except FileNotFoundError:
+            self._warn("Cannot find compiler %s in PATH." % compiler)
+
+        return self._filter_defines(result)
+
+    def _get_default_include_paths(self, compiler, std, language):
+        import subprocess
+        import re
+
+        result = list()
+
+        args = [compiler, '-x', language]
+        if std is not None:
+            args += ['-std=' + std]
+        args += ['-Wp,-v', '-E', '-']
+
+        try:
+            res = subprocess.Popen(
```

Comment:
Do you think we can reuse a [function from tools](https://github.com/niosus/EasyClangComplete/blob/master/plugin/tools.py#L687)?

Reply:
Ah, cool, definitely, I think so! One question: In the existing code I catch the potential `FileNotFoundError`; if I see it correctly, this is not done in the `run_command` utility. Shall we catch that exception in `run_command` or further up in the caller?

Reply:
Yeah, it just never occured yet :) Let's catch it there to be consistent.